Plugin/Sales/Block/Adminhtml/Order/View.php

```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * Taxjar_SalesTax
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Taxjar
+ * @package    Taxjar_SalesTax
+ * @copyright  Copyright (c) 2020 TaxJar. TaxJar is a trademark of TPS Unlimited, Inc. (http://www.taxjar.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace Taxjar\SalesTax\Plugin\Sales\Block\Adminhtml\Order;
+
+class View
+{
+    /**
+     * @var \Taxjar\SalesTax\Helper\Data
+     */
+    protected $_tjSalesTaxData;
+
+    /**
+     * @param \Taxjar\SalesTax\Helper\Data $tjSalesTaxData
+     */
+    public function __construct(\Taxjar\SalesTax\Helper\Data $tjSalesTaxData)
+    {
+        $this->_tjSalesTaxData = $tjSalesTaxData;
+    }
+
+    /**
+     * @param \Magento\Sales\Block\Adminhtml\Order\View $subject
+     */
+    public function beforeSetLayout(\Magento\Sales\Block\Adminhtml\Order\View $subject)
+    {
+        if ($this->_tjSalesTaxData->isSyncableOrder($subject->getOrder())) {
```

Comment:
It appears that this allows the sync button to be displayed even when transaction sync is disabled. This creates a scenario where the sync actually fails (but without throwing an exception) and results in this alert box:

![image](https://user-images.githubusercontent.com/11321890/143096216-0efeaa56-ad86-4eb1-8644-a93bb16b566c.png)

When transaction sync is disable in the scope of the order I believe that the button should not be displayed.

Reply:
This is a good find and I have fixed it, but while addressing it uncovered another scenario with similar logic to consider..

Do we want to display the TJ transaction sync date in the Sales Order info UI even if transaction sync is disabled?

Currently we do, but is that necessarily a good thing? I think that unlike this bug you've found here, where we obviously only want to show the "Sync to Taxjar" button (or include related JS) if transaction sync is enabled, it makes more sense if this sync-date partial is always included _if TaxJar is enabled_.

Take the case where transaction sync feature was enabled, but user no longer wishes to use it and disables it. That TJ sync date will still be displayed (i.e. the `.phtml` template/partial will be applied to the page) if the order has a `tj_salestax_sync_date` value stored on it, regardless of whether that feature is enabled. The argument could be made that because the data exists, it should be available to admin, but I'm not sure I agree.

If after installing the TJ SalesTax module, you decide to uninstall it, clear static content, recompile, and run upgrade, the `tj_salestax_sync_date` column will still exist on the `sales_order` table, and any existing values will remain. That data is still there, just hidden because TJ is "turned off" in M2. While that is partly due to the fact that our module doesn't include extension attributes properly for legacy reasons, I think it is also indicative of the fact that whether or not we included this UI template should depend on if our module is enabled, not the state of an order. The state of the order will then affect what we display within that UI template.

Reply:
This is a good find and I have fixed it, but while addressing it uncovered another scenario with similar logic to consider..

Do we want to display the TJ transaction sync date in the Sales Order info UI even if transaction sync is disabled?

Currently we do, but is that necessarily a good thing? I think unlike this(^) bug you've found here, where obviously we only want to show the "Sync to Taxjar" button (or include related JS) if transaction sync is enabled, it makes more sense to me if this sync-date partial is only included if transaction sync is enabled.

Take the case where transaction sync feature was enabled, but user no longer wishes to use it and so disables it. Regardless of whether transaction sync is enabled, the TJ sync date will still be displayed (i.e. the .phtml template/partial will be applied to the page) if the order has a tj_salestax_sync_date value stored on it.

The argument could be made that because the data exists, it should be visible to admin, but I'm not sure I agree. If after installing the TJ SalesTax module, you decide to uninstall it, clear static content, recompile, and run upgrade, the tj_salestax_sync_date column will still exist on the sales_order table, and any existing values will remain. That data is still there on the order, just hidden because TJ is "turned off" in M2. While that is partly due to the fact that our module doesn't include extension attributes properly for legacy reasons, I think it is also indicative of the fact that whether or not we include this UI template should first depend on if our module is enabled, then if a feature is configured, and finally the state of an order. The state of the order can then affect what we display within that UI template.

If we used extension attributes as recommended instead of storing our values directly on the Magento_Sales tables/models, would we still load an order's tj_salestax_sync_date value in our extension data when the TaxJar transaction sync feature flag is disabled?

What are your thoughts?

Reply:
I'll think about this more but for now I think its probably fine as is. If we do decide to make this change later we can open up a new card for it.